File: backend/app/schemas/ai.py
from __future__ import annotations

from pydantic import BaseModel, Field


class ChatSource(BaseModel):
    documentName: str | None = None
    title: str | None = None
    content: str | None = None
    score: float | None = None


class DifyChatFile(BaseModel):
    type: str
    transfer_method: str = "local_file"
    upload_file_id: str


class DifyUploadedFile(BaseModel):
    id: str
    name: str | None = None
    size: int | None = None
    extension: str | None = None
    mime_type: str | None = None
    type: str = "image"


class SendChatRequest(BaseModel):
    query: str = Field(min_length=1)
    conversation_id: str | None = None
    student_id: str | None = None
    major_code: str | None = None
    major_name: str | None = None
    intent: str | None = None
    inputs: dict | None = None
    files: list[DifyChatFile] = Field(default_factory=list)


class SendChatResponse(BaseModel):
    answer: str
    conversation_id: str
    sources: list[ChatSource] = []
    intent: str | None = None
    need_personal_data: bool = False


class ChatMessageItem(BaseModel):
    role: str
    content: str
    intent: str | None = None
    sources: list[ChatSource] = []
    need_personal_data: bool = False


class ChatSessionItem(BaseModel):
    id: str
    title: str | None = None
    time: str | None = None


class ChatHistoryResponse(BaseModel):
    conversation_id: str
    title: str | None = None
    messages: list[ChatMessageItem]


class ErrorCaseCreate(BaseModel):
    question: str
    wrong_answer: str | None = None
    expected_answer: str | None = None
    reason: str | None = None


class ErrorCaseStatusUpdate(BaseModel):
    status: str


class ErrorCaseItem(BaseModel):
    id: int
    question: str
    wrong_answer: str | None = None
    expected_answer: str | None = None
    reason: str | None = None
    status: str

File: frontend/src/types/ai.ts
export interface KnowledgeSource {
  title: string;
  datasetName?: string;
  documentName?: string;
  score?: number;
  content?: string;
}

export interface ChatMessage {
  id: string;
  role: 'user' | 'assistant';
  content: string;
  createdAt: string;
  loading?: boolean;
  sources?: KnowledgeSource[];
  intent?: string;
  statusSteps?: string[];
  currentStatus?: string;
}

export interface ChatSession {
  id: string;
  title?: string;
  time?: string;
}

export interface SendChatRequest {
  query: string;
  conversation_id?: string;
  student_id?: string;
  major_code?: string;
  major_name?: string;
  intent?: 'general_qa' | 'graduation_requirements' | 'graduation_audit' | 'course_advice';
  inputs?: Record<string, any>;
  files?: DifyChatFile[];
}

export interface DifyChatFile {
  type: string;
  transfer_method: 'local_file';
  upload_file_id: string;
}

export interface DifyUploadedFile {
  id: string;
  name?: string;
  size?: number;
  extension?: string;
  mime_type?: string;
  type: string;
}

export interface SendChatResponse {
  answer: string;
  conversation_id: string;
  sources?: KnowledgeSource[];
  intent?: string;
  need_personal_data?: boolean;
}

export interface ChatHistoryResponse {
  conversation_id: string;
  title?: string;
  messages: Array<{
    role: 'user' | 'assistant';
    content: string;
    intent?: string;
    sources?: KnowledgeSource[];
    need_personal_data?: boolean;
  }>;
}
